File: apps/web/src/backends/tauri/supervisor.ts
/**
 * The Rust supervisor, as the webview sees it — Phase 5 T4.
 *
 * `docs/findings/T3-rust-supervisor.md` §2 is the surface this describes, and
 * nothing here adds to it. Rust owns the two things a webview cannot do — spawn
 * `qemu-system-xtensa.exe`, hold the UART0 socket — and hands back **raw
 * bytes**. Everything above the bytes (the `@SESAME` parser, `encodeCommand()`,
 * the completion barrier, the capability record) is the TypeScript that already
 * exists and is already tested. That division is the whole of option C in
 * `docs/plans/phase-5-tauri-desktop-app.md` §4.
 *
 * ## Two channels, and the rule that makes them two
 *
 * ```text
 * uart   raw bytes  ->  SesameTelemetryParser        (the wire, and only the wire)
 * events JSON       ->  attempt / booted / diagnostic / exited / …
 * ```
 *
 * QEMU's own stdout and stderr arrive on the **event** channel as `diagnostic`
 * and never on the byte channel. That is not tidiness: a `log` event tagged
 * `provenance: 'observed'` carrying something the *emulator* said about itself
 * would be the laundering this project exists to refuse. T3 asserts the
 * separation against the real stream on every cycle; this side must not undo it
 * by folding diagnostics back into the parser.
 *
 * ## Why the transport is an interface
 *
 * `TauriSesameRobot` is held to `describeRobotContract`, and those fifteen cases
 * are Node code — C14 opens a real `node:http` listener — so they cannot run
 * inside a WebView2 window. Making the transport a parameter lets the *same*
 * robot class run in the suite against the *same* Rust supervisor, reached
 * through `--supervisor-stdio` instead of through IPC. See
 * `src-tauri/src/stdio.rs`, which states the two alternatives that were
 * rejected and why a fake would have been worthless here.
 *
 * ## No new dependency
 *
 * `@tauri-apps/api` is still not a dependency of this app. `withGlobalTauri` is
 * `true` in `src-tauri/tauri.conf.json`, so `window.__TAURI__.core` carries both
 * `invoke` and `Channel`, and that is what this uses — the same choice, for the
 * same reason, that `desktop/resource-report.ts` made in T2. Every access is
 * guarded rather than assumed, because this module is bundled into the browser
 * build too, where none of it exists.
 */

/**
 * What Rust reports about the origin: **facts, and only facts.**
 *
 * T3 §7 is explicit about what is deliberately absent here — `elided`, `board`,
 * `firmwareDeviations` and every capability boolean. Those are derived from the
 * image path by frozen, tested objects in `@sesame-lab/sesame-qemu`, and a
 * second hand-typed copy in Rust would be exactly the drift that lets the
 * packaged app claim something the web app does not. Rust carries the identity
 * across; the existing derivation decides what it means.
 */
export interface SupervisorOriginFacts {
  /** Always `'emulator'`. There is no branch in the crate producing another. */
  readonly kind: string;
  /**
   * `qemu-system-xtensa --version`, read out of the binary that was actually
   * spawned — not a constant. `null` if it would not answer.
   */
  readonly engine: string | null;
  /** The `-machine` this session booted. */
  readonly machine: string;
  /** Absolute path of the flash image, as resolved from the bundle. */
  readonly imagePath: string;
  /** Its file name — what `imageHasOledHook()` keys `oledFramebuffer` off. */
  readonly imageName: string;
  /** Absolute path of the emulator binary that was spawned. */
  readonly qemuPath: string;
}

/** One boot attempt, as Rust logged it. `ok: false` is ISSUE-20260823-022. */
export interface SupervisorBootAttempt {
  readonly attempt: number;
  readonly ok: boolean;
  readonly reason?: string;
  readonly ms: number;
}

/** What `spawn_emulator` answers with. */
export interface SupervisorSession {
  readonly pid: number;
  readonly port: number;
  readonly origin: SupervisorOriginFacts;
  readonly snapshot: boolean;
  /** The exact argv, so `snapshot=on` is auditable rather than trusted. */
  readonly args: readonly string[];
  /** Every attempt, failures included. */
  readonly attempts: readonly SupervisorBootAttempt[];
  readonly bootMs: number;
  readonly totalMs: number;
  readonly teardownEnforcedByJobObject: boolean;
  /** The UART0 write budget Rust enforces. 192; see `MAX_BATCH_BYTES`. */
  readonly maxWriteBytes: number;
}

/** Progress, out of band from the bytes. `type` is the tag. */
export type SupervisorEvent =
  | { readonly type: 'attempt'; readonly attempt: number; readonly of: number }
  | {
      readonly type: 'attemptFailed';
      readonly attempt: number;
      readonly of: number;
      readonly reason: string;
      readonly ms: number;
    }
  | {
      readonly type: 'booted';
      readonly attempt: number;
      readonly of: number;
      readonly ms: number;
      readonly pid: number;
      readonly port: number;
    }
  | { readonly type: 'guestPanic'; readonly text: string }
  /** QEMU's own stdout/stderr. **Not guest output.** Never parsed. */
  | { readonly type: 'diagnostic'; readonly stream: string; readonly text: string }
  | { readonly type: 'exited'; readonly code?: number }
  | { readonly type: 'stopped'; readonly pid: number; readonly confirmed: boolean };

/** What `stop_emulator` answers with. `wasRunning: false` is not an error. */
export interface SupervisorStopReport {
  readonly wasRunning: boolean;
  readonly pid?: number;
}

/** What `emulator_status` answers with. Cheap; safe to poll. */
export interface SupervisorStatusReport {
  readonly running: boolean;
  readonly pid?: number;
  readonly port?: number;
  /** Non-null means the guest died and every other field is last-known. */
  readonly guestPanic?: string;
}

/**
 * The knobs Rust exposes.
 *
 * **There are no path fields and that is the design** (T3 §3, deviation 2). A
 * webview that could hand in an arbitrary image could change what
 * `capabilitiesForImage()` reports about pixels it never observed. The paths
 * come from `resources::resolve` and nowhere else, which is how V7's property —
 * *the origin claim comes from the backend rather than the app asserting it* —
 * survives the move off HTTP.
 */
export interface SupervisorSpawnOptions {
  readonly bootAttempts?: number;
  readonly bootTimeoutMs?: number;
  readonly uartPort?: number;
  readonly snapshot?: boolean;
}

/**
 * The transport `TauriSesameRobot` sits on.
 *
 * Four methods, matching T3's four commands one for one. Anything that is not
 * *process lifetime* or *the socket* is deliberately absent: there is no
 * `command()` here, because encoding a command is `encodeCommand()`'s job and
 * fencing one is the barrier protocol's, and both live above this line.
 */
export interface EmulatorSupervisor {
  /**
   * Boot, and stream. `onBytes` receives whatever the socket carried, unframed
   * and in order — including everything buffered before the boot banner, which
   * is where `@SESAME hello` and the `rest` face `setup()` ends with live.
   */
  spawn(
    options: SupervisorSpawnOptions,
    onBytes: (chunk: ArrayBuffer | Uint8Array) => void,
    onEvent: (event: SupervisorEvent) => void,
  ): Promise<SupervisorSession>;
  /** Write raw bytes to UART0. Resolves with the count; does **not** fence. */
  send(bytes: Uint8Array): Promise<number>;
  stop(): Promise<SupervisorStopReport>;
  status(): Promise<SupervisorStatusReport>;
  /** Release the transport itself. A no-op over IPC; closes the pipe over stdio. */
  dispose?(): Promise<void>;
}

/** The shape `withGlobalTauri: true` installs on `window`. */
interface GlobalTauriCore {
  readonly invoke?: (cmd: string, args?: Record<string, unknown>) => Promise<unknown>;
  readonly Channel?: new <T>() => { onmessage: (message: T) => void };
}
interface GlobalTauri {
  readonly core?: GlobalTauriCore;
}

/** `window.__TAURI__.core`, or `null` outside the desktop shell. */
function core(scope: object = globalThis): GlobalTauriCore | null {
  const tauri = (scope as { __TAURI__?: GlobalTauri }).__TAURI__;
  const found = tauri?.core;
  if (found === undefined) return null;
  return typeof found.invoke === 'function' && typeof found.Channel === 'function' ? found : null;
}

/**
 * The supervisor over Tauri IPC, or `null` in a browser tab.
 *
 * `null` rather than a throwing stub: a browser has no emulator to supervise,
 * and answering with an object that fails on first use would put the discovery
 * one step further from the place that can explain it.
 */
export function tauriSupervisor(scope: object = globalThis): EmulatorSupervisor | null {
  const api = core(scope);
  if (api === null) return null;
  const invoke = api.invoke as NonNullable<GlobalTauriCore['invoke']>;
  const ChannelCtor = api.Channel as NonNullable<GlobalTauriCore['Channel']>;

  return {
    async spawn(options, onBytes, onEvent) {
      // `Response::new(Vec<u8>)` on the Rust side is an `InvokeResponseBody::Raw`,
      // so this channel delivers an ArrayBuffer rather than a JSON array of
      // 4,096 numbers per OLED frame. `Channel` stamps every message with an
      // index and reassembles in order — a reordered byte stream is a corrupted
      // one, and the parser downstream cannot tell.
      const uart = new ChannelCtor<ArrayBuffer>();
      uart.onmessage = (chunk) => onBytes(chunk);
      const events = new ChannelCtor<SupervisorEvent>();
      events.onmessage = (event) => onEvent(event);
      return (await invoke('spawn_emulator', { options, uart, events })) as SupervisorSession;
    },
    async send(bytes) {
      // A plain array, not a typed array: the IPC boundary serializes
      // arguments as JSON, and at most 192 bytes per write (T3 enforces it) is
      // small enough that the encoding cost is irrelevant. The *receive*
      // direction is where raw framing matters, and that is what the channel
      // above is for.
      return (await invoke('send_command', { bytes: Array.from(bytes) })) as number;
    },
    async stop() {
      return (await invoke('stop_emulator')) as SupervisorStopReport;
    },
    async status() {
      return (await invoke('emulator_status')) as SupervisorStatusReport;
    },
  };
}

File: src-tauri/src/supervisor.rs
//! The Tauri adapter over [`crate::qemu`] — Phase 5 T3's command surface.
//!
//! Thin on purpose. Everything that is hard — the process, the socket, the
//! retry loop, the job object — is in `qemu`, which knows nothing about Tauri
//! and can therefore be tested from a plain binary. This file resolves two
//! paths, turns two closures into two IPC channels, and holds the one live
//! session.
//!
//! ## The channel shape, and why there are two of them
//!
//! ```text
//! uart   : Channel<tauri::ipc::Response>   raw bytes, ArrayBuffer in JS
//! events : Channel<SupervisorEvent>        JSON progress, never bytes
//! ```
//!
//! The byte channel carries **the wire and nothing else**. Nothing is added to
//! it, nothing is removed from it, and no framing is applied — the
//! `SesameTelemetryParser` on the other side is proven chunk-invariant across
//! ~1,500 split offsets and must see exactly what the socket carried.
//!
//! That is also why QEMU's own stdout/stderr goes on the *event* channel as
//! `Diagnostic` and never on the byte channel. `session.ts` states the rule and
//! it is the sharpest honesty rule in the package: a log event claiming
//! `provenance: observed` for something the emulator said *about itself* would
//! be exactly the laundering the whole design exists to avoid.
//!
//! `tauri::ipc::Response::new(Vec<u8>)` produces an `InvokeResponseBody::Raw`,
//! which reaches JavaScript as an `ArrayBuffer` — a JSON array of 4,096 numbers
//! per OLED frame would be neither honest nor fast. Ordering is preserved by
//! `Channel` itself: every message carries an index and the JS side reassembles
//! in order, which matters because a reordered byte stream is a corrupted one.
//!
//! ## The paths are not parameters
//!
//! `spawn_emulator` takes no image path and no QEMU path. They come from
//! [`crate::resources::resolve`] and nowhere else — T2 §6.2 — and making them
//! un-overridable from the webview is the same property V7 chose the HTTP
//! adapter for: the app cannot assert what it is talking to. A frontend that
//! could hand in an arbitrary image could change what
//! `capabilitiesForImage()` reports about pixels it did not observe.

use std::sync::{Arc, Mutex};

use serde::{Deserialize, Serialize};
use tauri::ipc::{Channel, Response};
use tauri::{AppHandle, Manager};

use crate::qemu::{self, LaunchOptions, QemuError, Session, SessionInfo, SupervisorEvent};
use crate::resources;

/// The one live session, if there is one.
///
/// Managed as an `Arc` so a command can clone it out of the state and move it
/// into `spawn_blocking`: a boot is up to 12 attempts of up to 15 s and must
/// not run on the thread that draws the window.
#[derive(Default)]
pub struct Supervisor {
    session: Mutex<Option<Session>>,
}

impl Supervisor {
    /// Stop and forget the current session, if any. Idempotent.
    pub fn clear(&self) -> Option<u32> {
        let taken = self.session.lock().expect("session lock").take();
        taken.map(|session| {
            let pid = session.pid();
            session.stop();
            drop(session);
            pid
        })
    }
}

/// The knobs a caller may turn. Owned types only — an async command cannot
/// borrow its arguments.
///
/// Deliberately absent: `qemuPath` and `imagePath`. See the module docs.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct SpawnOptions {
    /// Boot attempts. Default 12 — Q2's measured budget, not a round number.
    pub boot_attempts: Option<u32>,
    /// Milliseconds for one attempt's banner. Default 15000.
    pub boot_timeout_ms: Option<u64>,
    /// UART0 TCP port. Default 0 = ask the OS for a free one.
    pub uart_port: Option<u16>,
    /// `-drive ...,snapshot=on`. Default true and there is no good reason to
    /// turn it off; see [`LaunchOptions::snapshot`]. Exposed only so a test can
    /// state that it deliberately wants a writable image.
    pub snapshot: Option<bool>,
}

impl SpawnOptions {
    fn apply(self, mut options: LaunchOptions) -> LaunchOptions {
        if let Some(attempts) = self.boot_attempts {
            options.boot_attempts = attempts.max(1);
        }
        if let Some(timeout) = self.boot_timeout_ms {
            options.boot_timeout_ms = timeout;
        }
        if let Some(port) = self.uart_port {
            options.uart_port = port;
        }
        if let Some(snapshot) = self.snapshot {
            options.snapshot = snapshot;
        }
        options
    }
}

/// What `stop_emulator` answers with.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StopReport {
    /// False when there was nothing to stop — not an error.
    pub was_running: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pid: Option<u32>,
}

/// What `emulator_status` answers with. Cheap; safe to poll.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusReport {
    pub running: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pid: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    /// The panic text, if the guest died under us.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub guest_panic: Option<String>,
}

/// Build [`LaunchOptions`] from the bundle. The only place paths come from.
pub fn options_from_resources(app: &AppHandle) -> Result<LaunchOptions, QemuError> {
    let qemu_path = resources::resolve(app, resources::QEMU_EXE)
        .map_err(|message| QemuError::Io { message })?;
    let image_path = resources::resolve(app, resources::FLASH_IMAGE)
        .map_err(|message| QemuError::Io { message })?;
    Ok(LaunchOptions::new(qemu_path, image_path))
}

fn supervisor(app: &AppHandle) -> Arc<Supervisor> {
    app.state::<Arc<Supervisor>>().inner().clone()
}

/// Boot the bundled QEMU and start streaming UART0.
///
/// Any session already running is stopped first, so a double-invoke leaks
/// nothing — that is one of the three repeat cases §7 of the plan names.
#[tauri::command]
pub async fn spawn_emulator(
    app: AppHandle,
    options: SpawnOptions,
    uart: Channel<Response>,
    events: Channel<SupervisorEvent>,
) -> Result<SessionInfo, QemuError> {
    let launch = options.apply(options_from_resources(&app)?);
    let state = supervisor(&app);
    state.clear();

    let byte_sink: qemu::ByteSink = Arc::new(move |bytes: &[u8]| {
        // A send failure means the webview is gone. There is nothing useful to
        // do about it here; the session is torn down by the exit hook.
        let _ = uart.send(Response::new(bytes.to_vec()));
    });
    let event_sink: qemu::EventSink = Arc::new(move |event| {
        let _ = events.send(event);
    });

    // Up to 12 × 15 s of blocking work. Off the UI thread, on Tauri's own
    // runtime — no second async runtime is introduced.
    let launched = tauri::async_runtime::spawn_blocking(move || {
        qemu::launch_with_retry(&launch, byte_sink, event_sink)
    })
    .await
    .map_err(|e| QemuError::Io {
        message: format!("the boot task did not finish: {e}"),
    })??;

    let (session, info) = launched;
    *state.session.lock().expect("session lock") = Some(session);
    Ok(info)
}

/// Kill QEMU and wait for the OS to confirm it.
#[tauri::command]
pub async fn stop_emulator(app: AppHandle) -> Result<StopReport, QemuError> {
    let state = supervisor(&app);
    let stopped = tauri::async_runtime::spawn_blocking(move || state.clear())
        .await
        .map_err(|e| QemuError::Io {
            message: format!("the stop task did not finish: {e}"),
        })?;
    Ok(StopReport {
        was_running: stopped.is_some(),
        pid: stopped,
    })
}

/// Write raw bytes to UART0.
///
/// **Raw.** T4 hands down the output of the existing `encodeCommand()` with the
/// existing `BARRIER_COMMAND` already appended; nothing here encodes, terminates
/// or frames anything. The one rule applied is the wire's own: at most
/// [`qemu::MAX_WRITE_BYTES`] per write, because arduino-esp32's 256-byte UART
/// ring buffer drains one character per `loop()` and overflows silently.
#[tauri::command]
pub async fn send_command(app: AppHandle, bytes: Vec<u8>) -> Result<usize, QemuError> {
    let state = supervisor(&app);
    tauri::async_runtime::spawn_blocking(move || {
        let guard = state.session.lock().expect("session lock");
        let session = guard.as_ref().ok_or_else(|| QemuError::NotConnected {
            message: "no emulator session; call spawn_emulator first".into(),
        })?;
        session.write(&bytes)
    })
    .await
    .map_err(|e| QemuError::Io {
        message: format!("the write task did not finish: {e}"),
    })?
}

/// Is anything running, and did it die?
#[tauri::command]
pub fn emulator_status(app: AppHandle) -> StatusReport {
    let state = supervisor(&app);
    let guard = state.session.lock().expect("session lock");
    match guard.as_ref() {
        None => StatusReport {
            running: false,
            pid: None,
            port: None,
            guest_panic: None,
        },
        Some(session) => StatusReport {
            running: session.alive(),
            pid: Some(session.pid()),
            port: Some(session.port()),
            guest_panic: session.panic_text(),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_defaults_survive_an_empty_options_object() {
        let base = LaunchOptions::new("q".into(), "i".into());
        let applied = SpawnOptions::default().apply(base.clone());
        assert_eq!(applied.boot_attempts, base.boot_attempts);
        assert_eq!(applied.boot_timeout_ms, base.boot_timeout_ms);
        assert!(applied.snapshot, "snapshot=on must survive; T2 §6.3");
    }

    #[test]
    fn zero_attempts_is_still_one_attempt() {
        let applied = SpawnOptions {
            boot_attempts: Some(0),
            ..Default::default()
        }
        .apply(LaunchOptions::new("q".into(), "i".into()));
        assert_eq!(applied.boot_attempts, 1);
    }

    /// The frontend cannot choose what it is talking to. If this ever gains a
    /// path field, `capabilitiesForImage()` becomes something the app asserts
    /// rather than something the backend reports.
    #[test]
    fn spawn_options_carry_no_paths() {
        let json = serde_json::json!({
            "qemuPath": "C:/evil/qemu.exe",
            "imagePath": "C:/evil/anything.flash.bin",
            "bootAttempts": 3
        });
        let parsed: SpawnOptions = serde_json::from_value(json).unwrap();
        assert_eq!(parsed.boot_attempts, Some(3));
        let applied = parsed.apply(LaunchOptions::new("real-qemu".into(), "real-image".into()));
        assert_eq!(applied.qemu_path, std::path::PathBuf::from("real-qemu"));
        assert_eq!(applied.image_path, std::path::PathBuf::from("real-image"));
    }

    #[test]
    fn clearing_an_empty_supervisor_is_not_an_error() {
        let supervisor = Supervisor::default();
        assert_eq!(supervisor.clear(), None);
        assert_eq!(supervisor.clear(), None);
    }
}
